File: mns-core/src/main/java/cz/fb/manaus/core/category/categorizer/SportCategorizer.java
package cz.fb.manaus.core.category.categorizer;

import cz.fb.manaus.core.MarketCategories;
import cz.fb.manaus.core.model.Market;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.Set;
import java.util.regex.Pattern;


@Component
final public class SportCategorizer extends AbstractDelegatingCategorizer {

    public static final Pattern AMERICAN_FOOTBALL = Pattern.compile("american\\s+football");
    public static final Pattern MOTOR_SPORT = Pattern.compile("motor\\s+sport");
    public static final Pattern ICE_HOCKEY = Pattern.compile("ice\\s+hockey");
    public static final String PREFIX = "sport_";

    public SportCategorizer() {
        super(PREFIX);
    }

    private Optional<String> getCategory(Market market) {
        var typeName = market.getEventType().getName().toLowerCase();
        if ("basketball".equals(typeName)) {
            return Optional.of(MarketCategories.BASKETBALL);
        } else if (AMERICAN_FOOTBALL.matcher(typeName).matches()) {
            return Optional.of(MarketCategories.AMERICAN_FOOTBALL);
        } else if (ICE_HOCKEY.matcher(typeName).matches()) {
            return Optional.of(MarketCategories.ICE_HOCKEY);
        } else if (MOTOR_SPORT.matcher(typeName).matches()) {
            return Optional.of(MarketCategories.MOTOR_SPORT);
        } else if ("volleyball".equals(typeName)) {
            return Optional.of(MarketCategories.VOLLEYBALL);
        } else if ("soccer".equals(typeName)) {
            return Optional.of(MarketCategories.SOCCER);
        } else if ("snooker".equals(typeName)) {
            return Optional.of(MarketCategories.SNOOKER);
        } else if ("cricket".equals(typeName)) {
            return Optional.of(MarketCategories.CRICKET);
        } else if ("handball".equals(typeName)) {
            return Optional.of(MarketCategories.HANDBALL);
        } else if (typeName.startsWith("greyhound")) {
            return Optional.of(MarketCategories.GREY_HOUNDS);
        } else if ("cycling".equals(typeName)) {
            return Optional.of(MarketCategories.CYCLING);
        } else if ("baseball".equals(typeName)) {
            return Optional.of(MarketCategories.BASEBALL);
        } else if ("golf".equals(typeName)) {
            return Optional.of(MarketCategories.GOLF);
        } else if ("tennis".equals(typeName)) {
            return Optional.of(MarketCategories.TENNIS);
        } else if (typeName.startsWith("horse")) {
            return Optional.of(MarketCategories.HORSES);
        } else if (typeName.startsWith("financial")) {
            return Optional.of(MarketCategories.FINANCIAL);
        } else if (typeName.startsWith("rugby")) {
            return Optional.of(MarketCategories.RUGBY);
        } else if (typeName.startsWith("politics")) {
            return Optional.of(MarketCategories.POLITICS);
        }
        return Optional.empty();
    }

    @Override
    public Set<String> getCategoryRaw(Market market) {
        var category = getCategory(market);
        return category.map(Set::of).orElse(Set.of());
    }
}

File: mns-core/src/main/java/cz/fb/manaus/core/category/categorizer/SportCategorizer.kt
package cz.fb.manaus.core.category.categorizer

import cz.fb.manaus.core.MarketCategories
import cz.fb.manaus.core.model.Market
import org.springframework.stereotype.Component
import java.util.regex.Pattern

const val SPORT_PREFIX = "sport_"

@Component
object SportCategorizer : AbstractDelegatingCategorizer(SPORT_PREFIX) {

    private val americanFootball = Pattern.compile("american\\s+football")!!
    private val motorSport = Pattern.compile("motor\\s+sport")!!
    private val iceHockey = Pattern.compile("ice\\s+hockey")!!

    private fun getCategory(market: Market): String? {
        val typeName = market.eventType.name.toLowerCase()
        when {
            "basketball" == typeName -> return MarketCategories.BASKETBALL
            americanFootball.matcher(typeName).matches() -> return MarketCategories.AMERICAN_FOOTBALL
            iceHockey.matcher(typeName).matches() -> return MarketCategories.ICE_HOCKEY
            motorSport.matcher(typeName).matches() -> return MarketCategories.MOTOR_SPORT
            "volleyball" == typeName -> return MarketCategories.VOLLEYBALL
            "soccer" == typeName -> return MarketCategories.SOCCER
            "esports" == typeName -> return MarketCategories.ESPORTS
            "table-tennis" == typeName -> return MarketCategories.TABLE_TENNIS
            "snooker" == typeName -> return MarketCategories.SNOOKER
            "cricket" == typeName -> return MarketCategories.CRICKET
            "handball" == typeName -> return MarketCategories.HANDBALL
            typeName.startsWith("greyhound") -> return MarketCategories.GREY_HOUNDS
            "cycling" == typeName -> return MarketCategories.CYCLING
            "baseball" == typeName -> return MarketCategories.BASEBALL
            "golf" == typeName -> return MarketCategories.GOLF
            "tennis" == typeName -> return MarketCategories.TENNIS
            typeName.startsWith("horse") -> return MarketCategories.HORSES
            typeName.startsWith("financial") -> return MarketCategories.FINANCIAL
            typeName.startsWith("rugby") -> return MarketCategories.RUGBY
            typeName.startsWith("politics") -> return MarketCategories.POLITICS
            else -> return null
        }
    }

    public override fun getCategoryRaw(market: Market): Set<String> {
        return when (val category = getCategory(market)) {
            null -> emptySet()
            else -> setOf(category)
        }
    }

}
